### PR TITLE
Allow parsed url to be passed down

### LINKED ISSUE #947: Avoid double qs parsing?
```jsx
  createServer((req, res) => {
    const { pathname, query } = parse(req.url, true)

    if (pathname === '/a') {
      app.render(req, res, '/b', query)
    } else if (pathname === '/b') {
      app.render(req, res, '/a', query)
    } else {
      handle(req, res)
    }
  })
```

I think `handle` should admit `query` as a param to avoid double-parsing?

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #947

### CODE REVIEW

**Comment by rauchg** [comment]
Something I thought of: we should re-parse qs if people didn't pass `true` to `parse` (ie: query is not an object)

**Comment by timneutkens** [comment]
@rauchg added 👍

File: server/index.js [hunk 1]
```diff
@@ -33,8 +33,12 @@ export default class Server {
   }
 
   getRequestHandler () {
-    return (req, res) => {
-      this.run(req, res)
+    return (req, res, url) => {
```
Comment (arunoda) [inline server/index.js]: I like to see `url` as `parsedUrl` here and everywhere else.

File: server/router.js [hunk 2]
```diff
@@ -14,16 +14,16 @@ export default class Router {
     this.routes.set(method, routes)
   }
 
-  match (req, res) {
+  match (req, res, url = parse(req.url)) {
```
Comment (arunoda) [inline server/router.js]: We don't need to parse the url here.
`.match()` is always called with a parsedUrl.

File: server/router.js [hunk 3]
```diff
@@ -14,16 +14,16 @@ export default class Router {
...
     if (!routes) return
 
-    const { pathname } = parse(req.url)
+    const { pathname } = url
     for (const r of routes) {
       const params = r.match(pathname)
       if (params) {
         return async () => {
-          return r.fn(req, res, params)
+          return r.fn(req, res, { ...params, url })
```
Comment (arunoda) [inline server/router.js]: I don't like keeping parsedUrl inside the `params`. Make it the 4th argument.

**Review by arunoda — CHANGES_REQUESTED** [review]

File: server/index.js (hunk 1 above)
Comment (timneutkens) [inline server/index.js]: Done 👍

File: server/router.js (hunk 2 above)
Comment (timneutkens) [inline server/router.js]: True, removed.

File: server/router.js (hunk 3 above)
Comment (timneutkens) [inline server/router.js]: Added as fourth argument.

**Review by arunoda — APPROVED** [review]

**Comment by arunoda** [comment]
Thanks @timneutkens

**Pull request merged**